### ISSUE TITLE
MacOS preprocessing performance

### ISSUE THREAD

**Issue by bkslasher** [fastbuild/fastbuild#535]
I've noticed that clean re-build of a large fully cached target is considerably slower on MacOS compared to Windows (`FBuild -cache -summary -clean Target`, 100% cache hits). It turns out that increasing worker threads count (`-j16`, 4-core CPU) improves performance by ~3 times in my case.
I've investigated this issue, and most of the time is spent during preprocessing. This is especially noticeable for large files (e.g. I have a Unity files with preprocessed sizes ~20 MB). Actual preprocessing to file (`clang ... -E -o out.txt`) in my case takes about 1.5 seconds, but FASTBuild's `ObjectNode::BuildPreprocessedOutput()` takes about 4 seconds. I think the problem is I/O bound by pipes that redirect clang's stdout to FASTBuild. As I understand MacOS' pipes use an internal buffer of up to 64K, so any writes beyond that cause a process to block.
A very simple fix that has worked for me is decreasing wait time inside Process.cpp `Process::ReadAllData()`:
```cpp
// no data available, but process is still going, so wait
// TODO:C Replace this sleep with event-based wait
Thread::Sleep( /*15*/ 1 );
```
But maybe there is a better way to fix it. Also I can't test this issue on Linux, but there may be the same problem there.

**Comment by ffulin** [comment]
Good find. 

On Windows, this issue was addressed by increasing the pipe buffer size. I think we'd want to do the same thing here. It looks like ```fntl``` with ```F_SETPIPE_SZ``` can be used to do this on Linux, but ```F_SETPIPE_SZ``` seems to not exist on mac and I found no alternative.

Assuming there is no way to set the buffer size for mac, a reduced timeout period along the lines of what you have tested seems like a good fix. As the lower sleep period will result in more context switches, it would be worth a quick test to see if a higher value (<15 but >1) could yield the same benefit while adding less context switches.

**Comment by bkslasher** [comment]
In my tests `Thread::Sleep(2)` looks like a sweet spot.

**Comment by ffulin** [comment]
Thanks. I built a version with Sleep( 2 ) and saw a halving of 100% cache hit build times, which is great. I also saw a ~5-10% improvement in general build performance (uncached or 0% cache hits).

I'll go ahead and make that change for OSX. I'll look into the pipe buffer change for Linux as well.

**Comment by ffulin** [comment]
I've submitted the changes (https://github.com/fastbuild/fastbuild/commit/f5a6ffa62e296e50a7c54824712528e2dfb05fc0). This includes:
 - setting 2ms sleep granularity for OSX
 - reducing sleep granularity to 8ms for Linux
 - setting pipe buffer sizes for Linux

I saw the same level of improvement for Linux as OSX.

These fixes will be in the next release (v0.98). I'll leave this issue open until that version is released.

Thanks for your help!
